torch/utils/pytree/__init__.py

```diff
@@ -0,0 +1,269 @@
+collection of Tensors and `tree_leaves` to get a flat list of all Tensors
+inside some nested collection. pytrees are helpful for implementing nested
+collection support for PyTorch APIs.
+"""
+
+from __future__ import annotations
+
+import os as _os
+import sys as _sys
+from dataclasses import dataclass as _dataclass
+from typing import (
+    Any as _Any,
+    Callable as _Callable,
+    Literal as _Literal,
+    TYPE_CHECKING as _TYPE_CHECKING,
+    TypeVar as _TypeVar,
+)
+from typing_extensions import ParamSpec as _ParamSpec, Self as _Self
+
+import torch.utils._pytree as python
+from torch._utils import classproperty as _classproperty
+from torch.utils._pytree import (  # these type aliases are identical in both implementations
+    FlattenFunc as FlattenFunc,
+    FlattenWithKeysFunc as FlattenWithKeysFunc,
+    FromDumpableContextFunc as FromDumpableContextFunc,
+    ToDumpableContextFunc as ToDumpableContextFunc,
+    UnflattenFunc as UnflattenFunc,
+)
+
+
+if _TYPE_CHECKING:
+    from types import ModuleType
+
+    from torch.utils._cxx_pytree import (  # noqa: TC004
+        PyTreeSpec as PyTreeSpec,
+        tree_all as tree_all,
+        tree_all_only as tree_all_only,
+        tree_any as tree_any,
+        tree_any_only as tree_any_only,
+        tree_flatten as tree_flatten,
+        tree_iter as tree_iter,
+        tree_leaves as tree_leaves,
+        tree_map as tree_map,
+        tree_map_ as tree_map_,
+        tree_map_only as tree_map_only,
+        tree_map_only_ as tree_map_only_,
+        tree_structure as tree_structure,
+        tree_unflatten as tree_unflatten,
+        treespec_pprint as treespec_pprint,
+    )
+
+
+__all__ = [
+    "PyTreeSpec",
+    "register_pytree_node",
+    "tree_flatten",
+    "tree_unflatten",
+    "tree_iter",
+    "tree_leaves",
+    "tree_structure",
+    "tree_map",
+    "tree_map_",
+    "tree_map_only",
+    "tree_map_only_",
+    "tree_all",
+    "tree_any",
+    "tree_all_only",
+    "tree_any_only",
+    "treespec_pprint",
+]
+
+
+if _TYPE_CHECKING:
+    __all__ += [
+        "FlattenFunc",
+        "UnflattenFunc",
+        "FlattenWithKeysFunc",
+        "ToDumpableContextFunc",
+        "FromDumpableContextFunc",
+    ]
+
+
+PYTORCH_USE_CXX_PYTREE: bool = _os.getenv("PYTORCH_USE_CXX_PYTREE", "0") not in {
+    "0",
+    "",
+}
+
+
+@_dataclass(frozen=True)
+class PyTreeImplementation:
+    """The underlying implementation for PyTree utilities."""
+
+    module: ModuleType = python
+    name: _Literal["python", "cxx"] = "python"
+
+    @_classproperty  # type: ignore[misc]
+    @classmethod
+    def python(cls) -> _Self:
+        """The Python implementation."""
+        return cls(module=python, name="python")
+
+    @_classproperty  # type: ignore[misc]
+    @classmethod
+    def cxx(cls) -> _Self:
+        """The C++ implementation."""
+        import torch.utils._cxx_pytree as cxx
+
+        return cls(module=cxx, name="cxx")
+
```

Comment:
The switching logic seems unnecessarily complicated. Why can't it just be:

```
if PYTORCH_USE_CXX_PYTREE:
   from torch.utils._cxx_pytree import <list_of_apis>
else:
   from torch.utils._pytree import <list_of_apis>
```
?

Reply:
I came to this first to allow the implementation module to be switched at runtime. If we doing direct imports, we cannot switch the implementation to another after the imported module is cached.

Reply:
> I came to this first to allow the implementation module to be switched at runtime.

If we do not want this. It can reduce some function call and `getattr` overheads.

Reply:
Why do we want the implementation to be swappable at runtime?

Also, if we do want this, we need to make c++ pytree and python pytree share the same treespec representation

Reply:
Updated.